packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
   }
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+   * @notice Executes a refundable meta-transaction on behalf of the signer.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param maxGasPrice The maximum gas price the user is willing to pay.
+   * @param gasLimit The gas limit for entire relayed transaction.
+   * @param metaGasLimit The gas limit for just the meta-transaction.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The return value of the meta-transaction execution.
+   */
+  function executeMetaTransactionWithRefund(
+    address destination,
+    uint256 value,
+    bytes calldata data,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit >= gasleft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
```

Comment:
An alternative to this would be specify they refunded gas price as `min(tx.gasPrice, maxGasPrice)`. The advantage would be to allow for relayers to use a higher gas price to get a transaction cleared, but if they do, they will only be refunded up to the amount the user agreed to ahead of time.

Reply:
If I am not mistaken, we could do the same for the overall gas limit. Assuming the user is not intending to cause an out-of-gas error (which is not impossible, but probably not something we are intending to support through this mechanism), this should not violate the user's wishes if a relayer wants to pay for some extra gas.